core/vm/evm.go

```diff
@@ -412,6 +421,21 @@ func (evm *EVM) create(caller ContractRef, codeAndHash *codeAndHash, gas uint64,
 	if !evm.Context.CanTransfer(evm.StateDB, caller.Address(), value) {
 		return nil, common.Address{}, gas, ErrInsufficientBalance
 	}
+	// For creation transaction (evm.depth == 0) EIP-3860 checks are done when calculating intrinsic gas
+	if evm.Config.HasEip3860() && evm.depth != 0 {
+		// Check whether the init code size has been exceeded.
+		len := len(codeAndHash.code)
+		if len > params.MaxInitCodeSize {
+			return nil, address, gas, ErrMaxInitCodeSizeExceeded
+		}
+		// Charge for init code size.
+		initCodeWordGas := toWordSize(uint64(len)) * params.InitCodeWordGas
+		if gas < initCodeWordGas {
+			return nil, address, gas, ErrCodeStoreOutOfGas
+		} else {
+			gas -= initCodeWordGas
+		}
```

Comment:
Is this correct? From the EIP: 

> For the CREATE and CREATE2 instructions charge an extra gas cost equaling to initcode_cost(initcode). This cost is deducted before the calculation of the resulting contract address and the execution of initcode. (Note that this means before or at the same time as the hashing cost is applied in CREATE2.) If the instruction fails due to the initcode length check of point 3, this cost is not deducted. 

As implemented here, if we do a create2 with 64K gas (all remaining). Then the 63/64ths rule will be applied, and it's scaled down to 63K gas. And only after that will this `initCodeWordGas` be removed.

From the wording in the eip, it sounds like it should be done much sooner, namely within the `gasCreate2` in `gas_table.go`. And if so, you could redefine `CREATE2` with a new gas calculator in the jump table.

Reply:
@holiman is right.